src/psyclone/tests/psyir/transformations/hoist_trans_test.py

```diff
@@ -0,0 +1,184 @@
+    BOOLEAN_TYPE, ArrayType
+from psyclone.psyir.transformations import HoistTrans, TransformationError
+from psyclone.tests.utilities import Compile
+
+
+# init
+
+def test_init():
+    '''Test a hoist transformation can be succesfully created.'''
+    hoist_trans = HoistTrans()
+    assert isinstance(hoist_trans, HoistTrans)
+
+
+# apply
+
+def test_apply(tmpdir):
+    '''Test the apply method moves the loop invariant assignment out of
+    the loop and places it immediately before the loop.
+
+    '''
+    code = (
+        "program test\n"
+        "  integer :: i\n"
+        "  real :: a\n"
+        "  do i=1,1\n"
+        "    a = 1.0\n"
+        "  end do\n"
+        "end program\n")
+    expected_code = (
+        "program test\n"
+        "  integer :: i\n"
+        "  real :: a\n\n"
+        "  a = 1.0\n"
+        "  do i = 1, 1, 1\n"
+        "  enddo\n\n"
+        "end program test\n")
+    reader = FortranReader()
+    psyir = reader.psyir_from_source(code)
+    assignment = psyir.walk(Assignment)[0]
+    hoist_trans = HoistTrans()
+    hoist_trans.apply(assignment)
+    writer = FortranWriter()
+    output = writer(psyir)
+    assert output == expected_code
+    assert Compile(tmpdir).string_compiles(output)
+
+
+def test_apply_validate():
+
+    '''Test the apply method calls the validate method.'''
+    hoist_trans = HoistTrans()
+    with pytest.raises(TransformationError) as info:
+        hoist_trans.apply(None)
+    assert ("The target of the HoistTrans transformation should be an "
+            "assignment, but found 'NoneType'." in str(info.value))
+
+
+# validate
+
+def test_validate_node():
+    '''Test the expected exception is raised if an invalid node is
+    supplied to the transformation.
+
+    '''
+    hoist_trans = HoistTrans()
+    with pytest.raises(TransformationError) as info:
+        hoist_trans.validate(None)
+    assert ("The target of the HoistTrans transformation should be an "
+            "assignment, but found 'NoneType'." in str(info.value))
+
+
+def test_validate_ancestor_loop():
+    '''Test the expected exception is raised if the supplied assignment is
+    not within a loop.
+
+    '''
+    assignment = Assignment.create(
+        Reference(DataSymbol("a", REAL_TYPE)), Literal("1.0", REAL_TYPE))
+    hoist_trans = HoistTrans()
+    with pytest.raises(TransformationError) as info:
+        hoist_trans.validate(assignment)
+    assert ("The supplied assignment node 'a = 1.0\n' should be within a "
+            "loop, but no loop was found." in str(info.value))
+
+
+def test_validate_direct_loop():
+    '''Test the expected exception is raised if the supplied assignment is
+    not the direct child of a loop (ignoring schedule).
+
+    '''
+    assignment = Assignment.create(
+        Reference(DataSymbol("a", REAL_TYPE)), Literal("1.0", REAL_TYPE))
+    condition = Literal("true", BOOLEAN_TYPE)
+    if_condition = IfBlock.create(condition, [assignment])
```

Comment:
As mentioned earlier, I can imagine we will want to support hoisting of a whole IfBlock. I think we would probably just have one HoistTrans that could be applied to (almost) any node type.

Reply:
Agreed. I added a note to the transformation and updated the issue as suggested earlier.